project3/coordinator/messenger.cpp

```diff
@@ -5,71 +5,62 @@
 #include "messenger.h"
 
 #include <chrono>
-#include <utility>
-
 #include <loguru.hpp>
+#include <utility>
 
 namespace coordinator {
 
-Messenger::Messenger(
-    std::shared_ptr<queue::Queue<MessageLog::Message>> msg_queue,
-    std::shared_ptr<MessageLog> msg_log,
-    std::shared_ptr<ConnectedParticipants::Participant> participant,
-    std::shared_ptr<ConnectedParticipants> connected_participants)
-    : msg_queue_(std::move(msg_queue)),
-      msg_log_(std::move(msg_log)),
-      connected_participants_(std::move(connected_participants)),
-      participant_(std::move(participant)) {}
+Messenger::Messenger(std::shared_ptr<MessageLog> msg_log,
+                     ConnectedParticipants::Participant participant)
+    : msg_log_(std::move(msg_log)), participant_(std::move(participant)) {}
+bool Messenger::operator==(const Messenger& messenger) const {
+  return participant_ == messenger.participant_;
+}
 
-bool Messenger::BroadcastMessage() {
-  // Pop next message from message queue.
-  MessageLog::Message msg = msg_queue_->Pop();
+ConnectedParticipants::Participant Messenger::GetParticipant() const {
+  return participant_;
+}
 
+bool Messenger::SerializeMessage(MessageLog::Message msg) {
   // Populate protobuf message.
   proto_msg_.set_message(msg.msg);
   proto_msg_.set_origin_id(msg.participant_id);
 
-  // Serialize the message.
+  // Serialize
   if (!wire_protocol::Serialize(proto_msg_, &outgoing_message_buffer_)) {
-    LOG_F(ERROR, "Failed to serialize message.");
     return false;
   }
-  // Send the message to each connected participant.
-  for (const ConnectedParticipants::Participant& participant :
-       connected_participants_->GetParticipants()) {
-    if (send(participant.sock_fd, outgoing_message_buffer_.data(),
-             outgoing_message_buffer_.size(), 0) < 0) {
-      LOG_F(ERROR, "Failed to send message.");
-      return false;
-    }
+  return true;
+}
+bool Messenger::SendMessage(MessageLog::Message msg, bool missed_msg) {
+  SerializeMessage(msg);
+  // mutex lock to ensure thread-safe socket sending.
+  std::lock_guard<std::mutex> guard(mutex_);
+  if (send(participant_.sock_fd, outgoing_message_buffer_.data(),
```

Comment:
Since you have a `proto_msg_` that's a class member, you probably want to grab the lock at the top of this function.